Added both ExpressionBody and BlockBody to the underlying BoundNode f…

333fred opened the pull request:
…or local functions, and exposed both to IOperation consumers.

<details><summary>Ask Mode template not completed</summary>

<!-- This template is not always required. If you aren't sure about whether it's needed or want help filling out the sections,
submit the pull request and then ask us for help. :) -->

### Customer scenario

IOperation analyzers cannot get full information in certain local function error scenarios.

### Bugs this fixes

Fixes https://github.com/dotnet/roslyn/issues/24650.

### Workarounds, if any

None.

### Risk

The main risk with this change is potentially reporting more errors if the additional bind doesn't properly discard errors, but I believe I handled this correctly. We're not breaking any public API, only adding 2 new fields to ILocalFunctionOperation.

### Performance impact

We're binding an expression we didn't before, but only in very specific error scenarios. This shouldn't affect most code.

### Is this a regression from a previous update?

No

### Root cause analysis

N/A

### How was the bug found?

Internal testing

### Test documentation updated?

N/A

</details>

333fred commented on src/Compilers/Core/Portable/Operations/ILocalFunctionOperation.cs:
```diff
@@ -23,10 +23,18 @@ public interface ILocalFunctionOperation : IOperation
-        /// Expression-bodied local functions are normalized here to be a single <see cref="IOperation"/>
-        /// inside of a block statement.
+        /// This will return the <see cref="BlockBody"/> if it exists, and the <see cref="ExpressionBody"/> if the <see cref="BlockBody"/> does not exist.
+        /// If both exist, this will return just the <see cref="BlockBody"/>, and you must use <see cref="ExpressionBody"/> to retrieve the expression body.
```
I chose to add two new properties here to mirror `IMethodBodyOperation`, and so that we could prevent a breaking change in the `Body` property by no longer returning anything for expression body form.

333fred replied:
@jinujoseph we should add this to the list for tomorrow's discussion.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ILocalFunctionOperation.cs:
```diff
@@ -23,10 +23,18 @@ public interface ILocalFunctionOperation : IOperation
+        /// <summary>
+        /// The block body of the local function, if it exists.
+        /// </summary>
+        IBlockOperation BlockBody { get; }
```
>IBlockOperation BlockBody { get; } [](start = 8, length = 34)

I think we should simply add one property```IBlockOperation IgnoredBody``` , similar to IWhileLoopOperation.

333fred replied:
Design Decision: Do this.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractRegionControlFlowPass.cs:
```diff
@@ -36,7 +36,9 @@ public override BoundNode VisitLambda(BoundLambda node)
         {
-            return VisitLocalFunctionOrLambda(node.Body);
+            VisitLocalFunctionOrLambda(node.BlockBody);
+            VisitLocalFunctionOrLambda(node.ExpressionBody);
```
>VisitLocalFunctionOrLambda(node.ExpressionBody); [](start = 12, length = 48)

I think we should not be changing this just because we have both in the node now. Unless we have a scenario for which we want the behavior to change. Do we have a scenario?

AlekseyTs replied:
Also, see how method bodies are handled in PreciseAbstractFlowPass.VisitMethodBodies. If we are going to visit both bodies, I would expect us to visit them in the same manner. I believe we do want to visit both for region analysis, so that WrittenInside and other sets are properly populated based on code in both bodies. We need to come up with test scenarios and find the right visitor that should be modified.  

---
In reply to: [177626096](https://github.com/dotnet/roslyn/pull/25782#discussion_r177626096) [](ancestors = 177626096)

AlekseyTs replied:
My recommendation is to not make any changes to control and data flow analysis along with the fix for #24650. Our primary goal for this change is to expose information through IOperation. We definitely should investigate if it makes sense to adjust control/data flow analysis for the situation of both bodies present, but we can do that later because this is a "preexisting condition". I suggest to open a follow-up issue for that.

---
In reply to: [177628448](https://github.com/dotnet/roslyn/pull/25782#discussion_r177628448) [](ancestors = 177628448,177626096)

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/25794](https://github.com/dotnet/roslyn/issues/25794) to follow up.

---
In reply to: [177815839](https://github.com/dotnet/roslyn/pull/25782#discussion_r177815839) [](ancestors = 177815839,177628448,177626096)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Statement.cs:
```diff
@@ -23,7 +23,7 @@ internal partial class BoundBadStatement
     partial class BoundLocalFunctionStatement
     {
-        protected override ImmutableArray<BoundNode> Children => ImmutableArray.Create<BoundNode>(this.Body);
+        protected override ImmutableArray<BoundNode> Children => ImmutableArray.Create<BoundNode>(this.BlockBody, this.ExpressionBody);
```
>protected override ImmutableArray<BoundNode> Children => ImmutableArray.Create<BoundNode>(this.BlockBody, this.ExpressionBody); [](start = 8, length = 127)

I think this code should be removed, we have support for local functions. See the comment in the base class above.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -492,50 +492,58 @@ private BoundStatement BindLocalFunctionStatement(LocalFunctionStatementSyntax n
-            if (block != null)
+            if (blockBody != null || expressionBody != null)
             {
                 localSymbol.ComputeReturnType();
```
> localSymbol.ComputeReturnType(); [](start = 15, length = 33)

It looks like this call is redundant, the ```GetDeclarationDiagnostics``` call is going to take care of this. I think we can remove this call and get rid of the ```if (blockBody != null || expressionBody != null)```.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -492,50 +492,58 @@ private BoundStatement BindLocalFunctionStatement(LocalFunctionStatementSyntax n
+                expressionBody = BindExpressionBodyAsBlock(node.ExpressionBody, expressionBodyDiagnostics);
             }
-            else
+            else if (node.Body == null && node.ExpressionBody == null)
```
>node.ExpressionBody == null [](start = 42, length = 27)

Is this condition ever  false?

333fred replied:
As written, it should be false if ExpressionBody is null but BlockBody isn't. We could remove the `node.ExpressionBody == null` part, but I'll rewrite this along your recommendations above so it'll be removed.

---
In reply to: [177632742](https://github.com/dotnet/roslyn/pull/25782#discussion_r177632742) [](ancestors = 177632742)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -492,50 +492,58 @@ private BoundStatement BindLocalFunctionStatement(LocalFunctionStatementSyntax n
+            if (node.ExpressionBody != null)
             {
-                block = BindExpressionBodyAsBlock(node.ExpressionBody, diagnostics);
+                expressionBodyDiagnostics = node.Body == null ? diagnostics : new DiagnosticBag();
```
>ode.Body == null [](start = 45, length = 16)

It looks like we are checking this condition and inlining it multiple times. It doesn't feel like this makes the code clearer, I would say quite the opposite. Consider refactoring into regular ```if``` statements. For example:
```
            BoundBlock blockBody = null;
            BoundBlock expressionBody = null;

            if (node.Body != null)
            {
                 blockBody = runAnalysis(BindEmbeddedBlock(node.Body, diagnostics), diagnostics);

                 if (node.ExpressionBody != null)
                 {
                     DIagnosticBag expressionBodyDiagnostics = new DiagnosticBag();
                     expressionBody = runAnalysis(BindExpressionBodyAsBlock(node.ExpressionBody, expressionBodyDiagnostics), expressionBodyDiagnostics);
                 }
            }
            else if (node.ExpressionBody != null)
            {
                 expressionBody = runAnalysis(BindExpressionBodyAsBlock(node.ExpressionBody, diagnostics), diagnostics);
            }
            else 
            {
                hasErrors = true;
                diagnostics.Add(ErrorCode.ERR_LocalFunctionMissingBody, localSymbol.Locations[0], localSymbol);
            }

            BoundBlock runAnalysis(BoundBlock block, DiagnosticBag blockDiagnostics)
            {
                if (block != null)
                {
                    // Have to do ControlFlowPass here because in MethodCompiler, we don't call this for synthed methods
                    // rather we go directly to LowerBodyOrInitializer, which skips over flow analysis (which is in CompileMethod)
                    // (the same thing - calling ControlFlowPass.Analyze in the lowering - is done for lambdas)
                    // It's a bit of code duplication, but refactoring would make things worse.
                    var endIsReachable = ControlFlowPass.Analyze(localSymbol.DeclaringCompilation, localSymbol, block, blockDiagnostics);
                    if (endIsReachable)
                    {
                        if (ImplicitReturnIsOkay(localSymbol))
                        {
                            block = FlowAnalysisPass.AppendImplicitReturn(block, localSymbol);
                        }
                        else
                        {
                            blockDiagnostics.Add(ErrorCode.ERR_ReturnExpected, localSymbol.Locations[0], localSymbol);
                        }
                    }
                }
                return block;
            }
```

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -5947,21 +5947,28 @@ internal abstract partial class BaseLocalFunctionStatement : Operation, ILocalFu
         /// </summary>
         public IMethodSymbol Symbol { get; }
-        protected abstract IBlockOperation BodyImpl { get; }
+        protected abstract IBlockOperation BlockBodyImpl { get; }
```
>BlockBodyImpl [](start = 43, length = 13)

Please get rid of the "Impl" pattern for this node, a merge of this change to dataflow branch shouldn't require any fix-ups, we should be able to simply accept code from the parent.

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -5977,29 +5984,34 @@ public override void Accept(OperationVisitor visitor)
             base(symbol, semanticModel, syntax, type, constantValue, isImplicit)
         {
-            BodyImpl = body;
+            BlockBodyImpl = blockBody;
```
>blockBody [](start = 28, length = 9)

The parent should be assigned right here

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -5977,29 +5984,34 @@ public override void Accept(OperationVisitor visitor)
         {
-            BodyImpl = body;
+            BlockBodyImpl = blockBody;
+            ExpressionBodyImpl = expressionBody;
```
>expressionBody [](start = 33, length = 14)

The parent should be assigned right here

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -5947,21 +5947,28 @@ internal abstract partial class BaseLocalFunctionStatement : Operation, ILocalFu
         /// </summary>
-        public IBlockOperation Body => Operation.SetParentOperation(BodyImpl, this);
+        public IBlockOperation Body => BlockBody ?? ExpressionBody;
+        public IBlockOperation BlockBody => Operation.SetParentOperation(BlockBodyImpl, this);
```
>Operation.SetParentOperation(BlockBodyImpl, this) [](start = 44, length = 49)

The lazy parent assignment should be moved to the lazy subclass

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs:
```diff
@@ -386,7 +386,7 @@ public override BoundNode VisitLambda(BoundLambda node)
 
                 public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
-                    => VisitClosure(node.Symbol.OriginalDefinition, node.Body);
+                    => VisitClosure(node.Symbol.OriginalDefinition, node.BlockBody ?? node.ExpressionBody);
```
>node.BlockBody ?? node.ExpressionBody [](start = 68, length = 37)

Consider adding calculated property "Body" to the node.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ILocalFunctionStatement.cs:
```diff
@@ -458,5 +471,85 @@ void M(int p)
+{
+    void M(int p)
+    {
+        /*<bind>*/object F() => new object(); { return new object(); }/*</bind>*/;
```
>new object(); [](start = 55, length = 13)

Consider using different expressions so that we could easily tell them apart. Applies to other added tests

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ILocalFunctionStatement.cs:
```diff
@@ -458,5 +471,85 @@ void M(int p)
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void TestInvalidLocalFunction_BlockAndExpressionBody()
```
>TestInvalidLocalFunction_BlockAndExpressionBody [](start = 20, length = 47)

It doesn't look like we test SemanticModel APIs for this scenario. Please make sure we have existing tests covering the case of both bodies present.

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [376759133](https://github.com/dotnet/roslyn/pull/25782#issuecomment-376759133) [](ancestors = 376759133)

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs:
```diff
@@ -1423,7 +1423,19 @@ public override void VisitLocalFunction(ILocalFunctionOperation operation)
+                {
+                    Visit(operation.Body);
+                }
+            }
```
>} [](start = 12, length = 1)

Consider asserting that peration.IgnoredBody is null in the else, to complete validation.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1396,7 +1396,15 @@ private CSharpSyntaxNode GetBindingRoot(CSharpSyntaxNode node)
                     case SyntaxKind.ArrowExpressionClause:
+                        // If this is an arrow expression on a local function statement, then our bindable root is actually our parent syntax as it's
+                        // a statement in a function. If this is returned directly in IOperation, we'll end up with a separate tree.
+                        if (current.ParentOrStructuredTriviaParent != null && current.ParentOrStructuredTriviaParent.Kind() == SyntaxKind.LocalFunctionStatement)
```
>ParentOrStructuredTriviaParent [](start = 36, length = 30)

I think ```.Parent``` would be better in this condition.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1396,7 +1396,15 @@ private CSharpSyntaxNode GetBindingRoot(CSharpSyntaxNode node)
                     case SyntaxKind.ArrowExpressionClause:
+                        // If this is an arrow expression on a local function statement, then our bindable root is actually our parent syntax as it's
+                        // a statement in a function. If this is returned directly in IOperation, we'll end up with a separate tree.
+                        if (current.ParentOrStructuredTriviaParent != null && current.ParentOrStructuredTriviaParent.Kind() == SyntaxKind.LocalFunctionStatement)
```
>if (current.ParentOrStructuredTriviaParent != null && current.ParentOrStructuredTriviaParent.Kind() == SyntaxKind.LocalFunctionStatement) [](start = 24, length = 137)

It is unfortunate that we have to do this. In SemanticModel we want to bind as little as possible, if we bubble up to the entire local function, we will be forced to bind entire block body, if it is present, and that block body could be quite big. From what I see things should work properly with the original implementation. It looks like ```MethodBodySemanticModel.Bind``` should be able to bind ArrowExpressionClause on its own and the problem is that that node is not getting attached to the BoundLocalFunctionStatement when we bind entire local function statement. I suspect that is happening because either incremental binder is not used to bind local function statement,  or the node is getting mangled by ```runAnalysis``` transformation in BindLocalFunctionStatement and, therefore, is not matched to what we already have in the cache. Also, the way block body is bound in BindLocalFunctionStatement prevents its reuse (see how Binder.BindMethodBody binds block so that it can be reused by SemanticModel). Anyway, all this looks like a "preexisting condition", we don't have to address all this in this PR, at least we should open an issue to follow up (with all this information).

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/25815](https://github.com/dotnet/roslyn/issues/25815) for followup.

---
In reply to: [177942610](https://github.com/dotnet/roslyn/pull/25782#discussion_r177942610) [](ancestors = 177942610)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1396,7 +1396,15 @@ private CSharpSyntaxNode GetBindingRoot(CSharpSyntaxNode node)
+                        // a statement in a function. If this is returned directly in IOperation, we'll end up with a separate tree.
+                        if (current.ParentOrStructuredTriviaParent != null && current.ParentOrStructuredTriviaParent.Kind() == SyntaxKind.LocalFunctionStatement)
+                        {
+                            return current.ParentOrStructuredTriviaParent;
```
>ParentOrStructuredTriviaParent [](start = 43, length = 30)

Parent?

AlekseyTs replied:
Or simply continue the loop?

---
In reply to: [177942626](https://github.com/dotnet/roslyn/pull/25782#discussion_r177942626) [](ancestors = 177942626)

333fred replied:
It seemed clearer to me to just return here, instead of inserting a continue. If we match a syntax here, we will return.

---
In reply to: [178117247](https://github.com/dotnet/roslyn/pull/25782#discussion_r178117247) [](ancestors = 178117247,177942626)

AlekseyTs replied:
> It seemed clearer to me to just return here, instead of inserting a continue. If we match a syntax here, we will return.

The code is structured so that we decide whether the node we are looking at is a binding root. If not, we continue, we don't try to make decision whether the parent is a good candidate instead.  

---
In reply to: [178128501](https://github.com/dotnet/roslyn/pull/25782#discussion_r178128501) [](ancestors = 178128501,178117247,177942626)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractRegionControlFlowPass.cs:
```diff
@@ -36,7 +36,8 @@ public override BoundNode VisitLambda(BoundLambda node)
         {
-            return VisitLocalFunctionOrLambda(node.Body);
+            VisitLocalFunctionOrLambda(node.Body);
+            return null;
```
>return null; [](start = 12, length = 12)

Consider completely reverting changes in this file.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs:
```diff
@@ -103,11 +106,9 @@ static void Main(string[] args)
-            Assert.Null(operation.Parent.Parent.Parent);
+            Assert.Equal(OperationKind.ExpressionStatement, objectCreationOperation.Parent.Kind);
+            Assert.Equal(OperationKind.Block, objectCreationOperation.Parent.Parent.Kind);
+            Assert.Equal(localFunctionOperation.IgnoredBody, objectCreationOperation.Parent.Parent);
```
>Equal [](start = 19, length = 5)

```Same```?

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
@dotnet/roslyn-compiler @dotnet/analyzer-ioperation I need another review please.

333fred commented:
Ping @dotnet/roslyn-compiler for a second review please.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -492,50 +492,61 @@ private BoundStatement BindLocalFunctionStatement(LocalFunctionStatementSyntax n
+                if (node.ExpressionBody != null)
+                {
+                    var expressionBodyDiagnostics = new DiagnosticBag();
+                    expressionBody = runAnalysis(BindExpressionBodyAsBlock(node.ExpressionBody, expressionBodyDiagnostics), expressionBodyDiagnostics);
```
Why drop these errors rather than reporting them?

333fred replied:
For now, I'm maintaining existing semantics here, which is to not report errors. I'll file a bug to follow up on this and consider reporting errors inside both bodies.

---
In reply to: [178338837](https://github.com/dotnet/roslyn/pull/25782#discussion_r178338837) [](ancestors = 178338837)

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/25845](https://github.com/dotnet/roslyn/issues/25845).

---
In reply to: [178372391](https://github.com/dotnet/roslyn/pull/25782#discussion_r178372391) [](ancestors = 178372391,178338837)

cston commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ILocalFunctionStatement.cs:
```diff
@@ -458,5 +459,111 @@ void M(int p)
+{
+    void M(int p)
+    {
+        /*<bind>*/object F() => new object(); { return null; }/*</bind>*/;
```
The `{ return null; }` is part of the outer method, not the local function.

333fred replied:
Right. I wanted to document this in an explicit IOperation test though so that we can confirm the behavior from the parsing in IOperation.

---
In reply to: [178340655](https://github.com/dotnet/roslyn/pull/25782#discussion_r178340655) [](ancestors = 178340655)

cston commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ILocalFunctionStatement.cs:
```diff
@@ -458,5 +459,111 @@ void M(int p)
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact, WorkItem(24650, "https://github.com/dotnet/roslyn/issues/24650")]
```
Does this test hit the issue?

333fred replied:
Yes. As part of fixing this bug, this expression would have returned an IOperation tree that wasn't part of the main tree.

---
In reply to: [178341100](https://github.com/dotnet/roslyn/pull/25782#discussion_r178341100) [](ancestors = 178341100)

333fred commented:
@dotnet-bot retest windows_debug_vs-integration_prtest please

333fred commented:
@dotnet-bot retest windows_release_vs-integration_prtest please
